snapcraft/tests/plugins/test_python.py

```diff
@@ -86,7 +211,93 @@ class Options:
         self.addCleanup(patcher.stop)
 
 
-class PythonPluginTestCase(BasePythonPluginTestCase):
+class PythonPluginPullTestCase(BasePythonPluginTestCase):
+
+    def test_with_unexisting_setup_file(self):
+        self.options.python_packages = ['dummy']
+        plugin = python.PythonPlugin('test-part', self.options,
+                                     self.project_options)
+        setup_directories(
+            plugin, self.options.python_version, setup_file=False)
+        with FakePip() as fake_pip:
+            plugin.pull()
+            # XXX Just a safeguard to make sure that the mock doesn't get out
+            # of sync.
+            self.assertThat(
+                fake_pip.mock_call.mock_calls[2][1][0],
```

Comment:
sorry, but `[2][1][0]` is even harder to read.

Reply:
Yeah this is tough to parse when trying to figure out what we're testing. Is this verifying that `download` isn't called without `'.'`? Or `install`? Or `wheel`?

Reply:
I think it was a mistake to add this for discussion. This is a [spy](http://xunitpatterns.com/Test%20Spy.html), and I'm not yet happy with any spy implementations I've tried.  
This is an unnecessary assertion, because if the setup.py doesn't exist and '.' is passed to the fake, it will raise an exception. So the fact that it doesn't fail is a good test. However, that's opaque and could lead to a fake that gets out of sync.

I've changed this to store in a boolean in a variable if '.' is passed to the fake. So it's more readable, but it's still opaque, it's not clear what leads to that variable being true by just reading at the test. I think that's better, but not perfect either.

And the third option would be to keep the `assert_has_calls` just as before, maybe replacing some already tested values with mock.ANY to reduce the burden of maintaining that. However, if we use `assert_has_calls` a lot, there's no reason to keep the fake.

So three options:
 * Rely on the fake raising an exception as the real would do.
 * If '.' is received as an argument, save that info in a variable.
 * Keep using `assert_has_calls`.

What do you think?